modules/core/arrow-extras-data/src/main/kotlin/arrow/data/fingertree/FingerTree.kt

```diff
@@ -249,6 +245,39 @@ sealed class FingerTree<T> : FingerTreeOf<T> {
 
     fun <A> fromArgs(vararg items: A): FingerTree<A> =
       fromList(items.toList())
+
+    fun <A, B> tailRecM(a: A, f: (A) -> Kind<ForFingerTree, Either<A, B>>): FingerTree<B> =
+      tailRecMHelper(empty(), f, f(a).fix())
+
+    private tailrec fun <A, B> tailRecMHelper(
```

Comment:
@pakoito could you also please explain the purpose of `tailRecM()` if you have some time? I tried to copy it from the `ListK` implementation, but I do not get the deeper meaning of it. Also, if the function `f()` returns a `Kind<F, Either<A, B >>`, with the first element being a `Either.Left`, the function will run forever?

Reply:
Sure! So, these data structures aren't stack-safe by default, meaning that if you write a recursive function that uses flatMap and handleErrorWith they'll end up throwing an StackOverflow because the compiler can't do tail-call optimization even when you say the function is `tailrec`.

These implementations use the heap and monadic behavior to prevent that. Hence, `tailrecM`.

Reply:
Ah, now I get it. Thanks for this detailed explanation!! 😄